ff/serving/provider/sql.go

```diff
@@ -0,0 +1,1225 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package provider
+
+import (
+	"database/sql"
+	"errors"
+	"fmt"
+	"github.com/google/uuid"
+	db "github.com/jackc/pgx/v4"
+	"strconv"
+	"strings"
+	"time"
+)
+
+func sanitize(ident string) string {
+	return db.Identifier{ident}.Sanitize()
+}
+
+//Change to struct
+type SQLOfflineStoreConfig struct {
+	Config        SerializedConfig
+	ConnectionURL string
+	Driver        string
+	ProviderType  Type
+	QueryImpl     OfflineTableQueries
+}
+
+type OfflineTableQueries interface {
+	setVariableBinding(b variableBindingStyle)
+	tableExists() string
+	resourceExists(tableName string) string
+	registerResources(db *sql.DB, tableName string, schema ResourceSchema, timestamp bool) error
+	primaryTableRegister(tableName string, sourceName string) string
+	primaryTableCreate(name string, columnString string) string
+	getColumnNames() string
+	getValueColumnTypes(tableName string) string
+	determineColumnType(valueType ValueType) (string, error)
+	materializationCreate(tableName string, resultName string) string
+	getTable() string
+	dropTable(tableName string) string
+	materializationIterateSegment(tableName string) string
+	newSQLOfflineTable(name string, columnType string) string
+	writeUpdate(table string) string
+	writeInserts(table string) string
+	writeExists(table string) string
+	createValuePlaceholderString(columns []TableColumn) string
+	trainingSetCreate(store *sqlOfflineStore, def TrainingSetDef, tableName string, labelName string) error
+	trainingRowSelect(columns string, trainingSetName string) string
+	castTableItemType(v interface{}, t interface{}) interface{}
+	getValueColumnType(t *sql.ColumnType) interface{}
+	numRows(n interface{}) (int64, error)
+	transformationCreate(name string, query string) string
+}
+
+type sqlOfflineStore struct {
+	db     *sql.DB
+	parent SQLOfflineStoreConfig
+	query  OfflineTableQueries
+	BaseProvider
+}
+
+// NewPostgresOfflineStore creates a connection to a postgres database
+// and initializes a table to track currently active Resource tables.
+func NewSQLOfflineStore(config SQLOfflineStoreConfig) (*sqlOfflineStore, error) {
+	url := config.ConnectionURL
+	db, err := sql.Open(config.Driver, url)
+	if err != nil {
+		return nil, err
+	}
+
+	return &sqlOfflineStore{
+		db:     db,
+		parent: config,
+		query:  config.QueryImpl,
+		BaseProvider: BaseProvider{
+			ProviderType:   config.ProviderType,
+			ProviderConfig: config.Config,
+		},
+	}, nil
+}
+
+func (store *sqlOfflineStore) getResourceTableName(id ResourceID) string {
+	var idType string
+	if id.Type == Feature {
+		idType = "feature"
+	} else {
+		idType = "label"
+	}
+	return fmt.Sprintf("featureform_resource_%s_%s_%s", idType, id.Name, id.Variant)
```

Comment:
what happens if name or variant has an underscore then:
name = abc_def
variant = ghi

is the same as
name = abc
variant = def_ghi

Reply:
Good point